pkg/registry/apis/provisioning/repository_controller.go

```diff
@@ -215,114 +198,119 @@ func (rc *RepositoryController) processNextWorkItem(ctx context.Context) bool {
-		logger.Info("handle repository deletion")
-		cfg, ok := item.obj.(*provisioning.Repository)
-		if !ok {
-			return errors.New("object is not a repository")
-		}
-
-		repo, err := rc.repoGetter.AsRepository(ctx, cfg)
-		if err != nil {
-			return fmt.Errorf("unable to create repository from object: %w", err)
-		}
-
-		if hooks, ok := repo.(repository.RepositoryHooks); !ok {
-			err := hooks.OnDelete(ctx)
-			if err != nil {
-				return fmt.Errorf("handle repository delete: %w", err)
-			}
-		}
-
-		parser, err := rc.parsers.GetParser(ctx, repo)
-		if err != nil {
-			return fmt.Errorf("failed to get parser for %s: %w", repo.Config().Name, err)
-		}
-
-		id, err := rc.identities.WorkerIdentity(ctx, namespace)
-		if err != nil {
-			return err
-		}
-		ctx = identity.WithRequester(ctx, id)
-
-		replicator, err := jobs.NewSyncer(repo, rc.resourceLister, parser)
-		if err != nil {
-			return fmt.Errorf("error creating replicator")
-		}
-
-		if err := replicator.Unsync(ctx); err != nil {
-			return fmt.Errorf("unsync repository: %w", err)
-		}
-
-		return nil
-	}
-
-	cachedRepo, err := rc.repoLister.Repositories(namespace).Get(name)
+	obj, err := rc.repoLister.Repositories(namespace).Get(name)
 	switch {
 	case apierrors.IsNotFound(err):
 		return errors.New("repository not found in cache")
 	case err != nil:
 		return err
 	}
 
-	id, err := rc.identities.WorkerIdentity(ctx, cachedRepo.Namespace)
+	healthAge := time.Since(time.UnixMilli(obj.Status.Health.Checked))
+	hasSpecChanged := obj.Generation != obj.Status.ObservedGeneration
+	if !hasSpecChanged && obj.DeletionTimestamp == nil && (healthAge < time.Hour*4) {
+		logger.Info("repository spec unchanged")
+		return nil
+	}
+
+	ctx := context.Background()
+	id, err := rc.identities.WorkerIdentity(ctx, namespace)
 	if err != nil {
 		return err
 	}
 	ctx = identity.WithRequester(ctx, id)
-	logger = logger.With("repository", cachedRepo.Name, "namespace", cachedRepo.Namespace)
-	ctx = logging.Context(ctx, logger)
+	logger = logger.WithContext(ctx)
 
-	repo, err := rc.repoGetter.AsRepository(ctx, cachedRepo)
+	repo, err := rc.repoGetter.AsRepository(ctx, obj)
 	if err != nil {
 		return fmt.Errorf("unable to create repository from configuration: %w", err)
 	}
 
-	hasSpecChanged := cachedRepo.Generation != cachedRepo.Status.ObservedGeneration
-	if !hasSpecChanged {
-		logger.Info("repository spec unchanged")
-		return nil
-	}
+	// Safe to edit the repository from here
+	obj = obj.DeepCopy()
+	hooks, _ := repo.(repository.RepositoryHooks)
+	status := &obj.Status
+	var sync *provisioning.SyncOptions
 
-	logger.Info("repository spec changed", "previous", cachedRepo.Status.ObservedGeneration, "current", cachedRepo.Generation)
+	switch {
+	// Delete (note this switch does not fallthrough to the health check)
+	case obj.DeletionTimestamp != nil:
+		logger.Info("handle repository delete")
+		if hooks != nil {
+			// FIXME... this should be converted to a "remove-webhook" finalizer
```

Comment:
```suggestion
			// FIXME... this should be converted to a "remove-webhook" finalizer
```

Not fully, it's a `remove-repository` as the webhook is behind the interface :)

Reply:
It is... but should it be?  EG -- when we actually add the webhook, we can add a finalizer to remove it "remove-github-webhook", and then we can have an explicit function to support that, and keep it independent from the